src/core/tokenizer.c

```diff
@@ -149,31 +154,83 @@ tokenset_resolve(tokenset_t *tset, nxs_index_t *idx, bool stage)
 	}
 }
 
+
+void		char_from_uchar_slice(char *, UChar* , int32_t , int32_t);
+void
+char_from_uchar_slice(char* buf, UChar* str, int32_t start, int32_t end)
+{
+	UChar   savedEndChar;
+
+	savedEndChar = str[end];
+	str[end] = 0;
+	u_austrcpy(buf, str+start);
+	str[end] = savedEndChar;
+}
+
+
 /*
- * tokenize: split the text into tokens separated by a whitespace
- * and pass the token through the filter pipeline.
+ * tokenize: Uses ICU segmentation UBRK_WORD
+ * See: https://unicode.org/reports/tr29/
  */
 tokenset_t *
-tokenize(filter_pipeline_t *fp, const char *text, size_t text_len __unused)
+tokenize(filter_pipeline_t *fp, const char *text, size_t text_len, const char *lang)
 {
-	const char *sep = ",.;:| \t\n";
-	char *content, *val, *brk;
-	tokenset_t *tset;
+	UBreakIterator *it_token;
+	UChar       *utext = NULL;
+	UErrorCode     ec = U_ZERO_ERROR;
+	int32_t	end;
+	int32_t	start;
+	tokenset_t	*tset;
+	int32_t	ulen;
 
 	if ((tset = tokenset_create()) == NULL) {
 		return NULL;
 	}
-	if ((content = strdup(text)) == NULL) {
+
+	ulen = roundup2((text_len + 1) * 3, 64);
```

Comment:
This will work, but I don't follow the calculation. Shouldn't you `malloc()` the following: `(text_len + 1) * 2`? Because UTF-8 to UTF-16 conversion can require twice as many bytes (incl. NUL), but no more. However, UTF-16 to UTF-8 conversion is likely result in more than half of the units, therefore reverse conversion calculation is different.

It can be pretty confusing.. I think we should introduce something like `utf8_calc_buflen()` and `utf16_calc_buflen()` functions in `utf8.c` and have these calculations in one place.

Reply:
I think it will be better if we could put these calculations and mem allocation inside the "utf8_to_utf16", or is that wrong?